### DISCUSSION TITLE
Suggestion: Text string tokens

### DISCUSSION THREAD

**Discussion by Matthew-Wise** [leekelleher/umbraco-contentment#70] [Idea]
I was wondering what you think extending the text string data type to have a token selector that can then be parsed out.

e.g

items  X  would be input as -> items [#count] instead of relying on the user inputting this have it selectable in some form.

In the case above it would then allow X items to be possible in a translated situation.

Happy to put the work in of course, just not sure on a UI. I see the tokens maybe been a DataListSource?

**Comment by leekelleher** [comment]
Hiya @Matthew-Wise!

I'm not fully following how the feature would work.

My thinking so far... with the [Text Input](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/text-input.md) editor, you'd enter text along with a token, from your example `[#count]`.  But what is `count`? Where does that come from? What's the context?

_I'm not sure how it relates to `IDataListSource` yet, but that may be clearer with discussion._

**Reply by Matthew-Wise** [reply]
count could be anything it's more of a placeholder. In case of count it could be how many items are in a basket for example.

may not be the best UI but hopefully shows what I mean

![image](https://user-images.githubusercontent.com/6782865/109813834-c2df9980-7c25-11eb-97db-b2379f238636.png)

So clicking on count or token x would add that placeholder which in code you would replace with the actual value in the same way that magic strings (should) in Umbraco forms - https://our.umbraco.com/documentation/add-ons/umbracoforms/developer/Magic-Strings/

**Reply by leekelleher** [reply]
@Matthew-Wise OK, gotcha. 👍 

The question to myself, is this something I need on my own client projects / want to develop / want to support / etc... and the honest answer is no, not really... and that's more about me than poo-poo'ing the idea. I like the idea, it has potential!

Some thoughts I'd had...

For the UI, thinking about Vendr's dictionary insert button for their textbox fields.

![vendr-textbox](https://user-images.githubusercontent.com/209066/109830045-0510d700-7c36-11eb-9e15-eaec6f04b8ea.jpg)

It'd open an overlay, with all the available (pre-configured) tokens, (even all the magic strings). The user would select one, the token is added to the textstring field.  Save & publish, all good.

The value-converter would scan the value for tokens, matching any with the pre-configured ones... if found, replace the token.

In terms of configuration, it could be a provider-based model - some interface comprising of the token's name/alias and a corresponding method to replace/transform the token. All of the magic string ones (e.g. Request, Session, Cookie, et al) could come as default.  Hmmm. 🤔

Sounds like a good one to hack on and explore, but like I say, not something I'm (currently) wanting for Contentment itself.

**Comment by leekelleher** [comment]
This idea has made me remember the Shortcodes package I did years ago. 🤔
<https://our.umbraco.com/packages/developer-tools/shortcodes/>
<https://github.com/leekelleher/umbraco-shortcodes>

**Reply by leekelleher** [reply]
💡  Once the ideas start flowing...

Thinking about the shortcodes gets me thinking about [Triple Tags](https://en.wikipedia.org/wiki/Tag_(metadata)#Triple_tags), something that was popular back with Flickr (in the day). The format is like this...

    [namespace]:[predicate]=[value]

So say for getting the value from `Session["count"]` could be reflected as a triple tag...

    [umb:session=count]

or forgetting triple tags, could be in the UDI (URI) schema? e.g. `umb://session/count`?

Of course, `[#count]` is more concise, but the prefixes (for the other [magic strings](https://our.umbraco.com/documentation/add-ons/umbracoforms/developer/Magic-Strings/)), do limit their scope. 🤔

_Anyway, just brain dumping._ 🧠

**Reply by leekelleher** [reply]
Another follow up... _(I'll stop spamming shortly.)_

I wrote a blog post last year about 
[using ClientDependency Filters to manipulate HTML](https://leekelleher.com/2020/01/clientdependency-filters-manipulate-html/), which could be hooked into to scan & parse any tokens/shortcodes. 🤔

The beauty of doing this at module Filter level is that any tokens in the markup could be replaced/transformed. It wouldn't be isolated to an editor's value-converter.  It'd open it up for other areas, such as a TinyMCE plugin.

_(OK, I'll stop now.)_

**Comment by leekelleher** [comment]
@Matthew-Wise Just spotted this in the package repo on Our... <https://our.umbraco.com/packages/backoffice-extensions/editor-with-replaceable-tokens/>. _(I've never tried it, think it's for v7.11)_

_(Also, worried that I scared you off with my previous responses?)_

**Reply by Matthew-Wise** [reply]
I had not seen it :D and is a better solution that I created for a client in v8 👍.

Scared off no just not sure where to go next with the idea 🤔
